libmamba/include/mamba/core/logging.hpp

```diff
@@ -0,0 +1,1183 @@
+            thread-safety requirements, @see `mamba::logging::LogHandler`.
+        */
+        auto flush_logs(std::optional<log_source> source = {}) -> void;
+
+
+        /** Set a flush threshold, a logging level for which log records pushed in `log` will
+            trigger a flush of logging sinks.
+
+            If a log handler is registered, this function calls `AnyLogHandler::set_flush_threshold`
+            with the same argument, otherwise this function will do nothing.
+            @see `mamba::logging:LogHandler` and @see `mamba::logging::AnyLogHandler` for details.
+
+            The flush threshold is the logging level for which if a log record is pushed
+            and has this logging level or higher, the implementation must flush it's sinks
+            immediately.
+
+            If the specified level is "all", the implementation must flush at every call to
+            `mamba::logging::log`.
+
+            The implementation is also free to flush at any other time but must guarantee
+            the flush is done after then end of this operation.
+
+            This call is thread safe as long as the log handler implementation fulfills the
+            thread-safety requirements, @see `mamba::logging::LogHandler`.
+
+            @see `mamba::logging::log`
+        */
+        auto set_flush_threshold(log_level threshold_level) -> void;
+
+        ///////////////////////////////////////////////////////
+        // MIGHT DISAPPEAR SOON
+        class MessageLogger
+        {
+        public:
+
+            MessageLogger(
+                log_level level,
+                std::source_location location = std::source_location::current()
+            );
+            ~MessageLogger();
+
+            std::stringstream& stream()
+            {
+                // Keep this implementation inline for performance reasons.
+                return m_stream;
+            }
+
+            static void activate_buffer();
+            static void deactivate_buffer();
+            static void print_buffer(std::ostream& ostream);
+
+        private:
+
+            log_level m_level;
+            std::stringstream m_stream;
+            std::source_location m_location;
+
+            static void emit(LogRecord log_record);
+        };
+
+        namespace details
+        {
+            // NOTE: this looks complicated because it's a workaround for `std::vector`
+            // implementations which are not `constexpr` (required by c++20), we defer the vector
+            // creation to the moment it's needed. Constexpr constructor is required for a type
+            // which is usable in a `constinit` definition, which is required to avoid the
+            // static-initialization-fiasco (at least for initialization).
+            // TODO: once homebrew stl impl has `constexpr` vector, replace all this by just `using
+            // MessageLoggerBuffer = vector<LogRecord>;`
+            struct MessageLoggerBuffer
```

Comment:
I guess this will disappear with MessageLogger?

Reply:
Yes but right now I cannot remove it without breaking something (I suspect) so I won't change it in this PR.
But yeah, if we remove MessageLogger, this whole goes away with it.